
**July 21st - 27th 2021**

# For [Course materials and detailed timetable visit here](https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2021/)


# **Note:** You are expected to be familiar with Rstudio, RMarkdown documents and Bioconductor. If you are not then you are advised to work through the material  (below) **before** attending Summer School
- [Introduction to Bioconductor](Introduction/bioc-intro.html).  
- [Bioconductor answers](Introduction/bioc-intro.solutions.html).  
  + [source](Introduction/bioc-intro.Rmd)
  + [Source with solutions](Introduction/bioc-intro.solutions.Rmd). 
  
# Day 0

**6.30pm SOCIAL: Online meet-up & Quiz ** 

# Day 1

**Please note 9.00 Start**
- 09:00 - 09:55; Welcome (Mark) & Introduction to Functional Genomics (Rory)
- 09:55 - 12:45; Data Processing for Next Generation Sequencing ( & )
    + Introduction to next generation sequencing 
    + A brief Introduction to genomic file formats 
    + Quality control and artefact removal 
    + Practical - Learn to use FastQC and Cutadapt on a sample dataset
    + Short Read alignment and quality control 
    + Practical - Alignment to reference genomes using BWA and STAR
    
- 12:45 - 13:45; LUNCH BREAK

- 13:45 - 17:15; Bulk RNAseq
    + Introduction to RNAseq Analysis in R
    + Quantification of Gene Expression with Salmon
    + Initial exploration of RNA-seq data 
    
# Day 2

- 09:30 - 17:00; Bulk RNAseq (continued)  
    + Statistical Analysis of Bulk RNAseq Data 
    + Experimental Design of Bulk RNAseq studies 
    + Differential Expression for RNA-seq  
    
- 12:30 - 13:30; LUNCH

# Day 3

- 09:30 - 12:30; Bulk RNAseq (continued) 
    + Annotation and Visualisation of RNA-seq results 
    + Gene-set testing 

- 12:30 - 13:30; LUNCH BREAK

- 13:30 - 17:30; Single Cell RNAseq (Stephane, Kasia, Ash)
    + 13:30 opening <!-- Stephane -->
    + 13:30 - 14:15 Introduction <!-- []() --> <!-- Kania - lecture -->
    + 14:15 - 14:30 Preamble: data set and workflow <!-- []() --> <!-- Stephane - 'lecture' -->
    + 14:30 - 15:55 library structure, cellranger for alignment and cell calling, preprocessing <!-- []() --> <!-- Ash -->
    + 15:55 - 16:05 **10 min break**
    + 16:05 - 17:30 normalisation <!-- []() --> <!-- Stephane -->

    
**6pm SOCIAL: Zoom Talk: “Collaborating with a Bioinformatician by Rory Stark”. **

# ---- Weekend (Machines will be available if you want to get a bit more time on them) ----

# Day 4

- 09:30 - 12:30; Single Cell RNAseq (cont.) (Zeynep, Abbi)
    + 09:30 - 10:55 dimensionality reduction, identification of confounding factors, feature selection <!-- []() --> <!-- Zeynep -->
    + 10:55 - 11:05 **10 min break**
    + 11:05 - 12:30 batch correction and data set integration <!-- []() --> <!-- Abbi -->

- 12:30 - 13:30; LUNCH BREAK

- 13:30 - 17:30; Single Cell RNAseq (cont.) (Stephane, Zeynep)
    + 13:30 - 14:30 clustering <!-- []() --> <!-- Stephane -->
    + 14:30 - 15:25 cluster marker genes <!-- []() --> <!-- Zeynep -->
    + 15:25 - 15:35 **10 min break**
    + 15:35 - 16:30 differential expression and abundance between conditions <!-- []() --> <!-- Stephane -->
    + 16:30 - 17:30 trajectories <!-- []() --> <!-- Zeynep -->
    + 17:30 closing <!-- Stephane -->
    
# Day 5 

- 09:30 - 17:00; ChIPseq (Joanna & Shoko)
    + Introduction to ChIP-seq 
    + Introduction to Peak Calling 
    + Differential binding analysis 
    + Quality control methods for ChIP-seq 
    + Downstream analysis of ChIP-seq 
    + Introduction to (epi)Genomic data integration
    + Chromatin accessibility and nucleosome occupancy with ATAC-seq 
    + Useful software utilities for the analysis of genomic data 

- 12:30 - 13:30; LUNCH BREAK

** End of CRUK CC Bioinformatics Summer School 2021**
-------------------------------------------------------------
